drivers: lora: sx1276: support more board designs

mniestroj opened the pull request:
This MR implements several new device-tree properties that will allow to configure sx1276 support on various board designs. Those are basically antenna selection pins and crystal enable pin.

Additionally one Kconfig option (power amplifier output selection) has been converted to device-tree enum string, as it is related strictly to device (PCB) design.

mniestroj committed:
drivers: lora: sx1276: support RFI/RFO/PA_BOOST antenna selection

There are several antenna path designs on SX1276 compatible boards in
the wild. B-L072Z-LRWAN1 board has dedicated enable pins for RFI, RFO
and PA_BOOST. This is exactly what this patch allows to
configure. Second variant of antenna selection is done with a single
GPIO pin, which controls RF SPDT switches (input or output). This is
also supported, when either 'rfo-enable-gpios' or
'pa-boost-enable-gpios' property is provided alone (RFO/PA_BOOST is
selected only when transmitting, so there is no need for explicit
'rfi-enable-gpios' configuration).

Drop requirement for 'power-amplifier-output' DT property when there is
either 'rfo-enable-gpios' or 'pa-boost-enable-gpios' configured. Fail
using BUILD_ASSERT() when neither is specified.

Make the SX1276SetAntSw() logic similar to loramac-node examples
implementation, so RFO/PA_BOOST is enabled only in
RFLR_OPMODE_TRANSMITTER.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj committed:
drivers: lora: sx1276: support antenna power enable

RF SPDT switches used for RX/TX selection have often power enable pin
connected to MCU's GPIO, so it is possible to disable it to save power
when not doing any data transfers. Add 'antenna-enable-gpios' property
to support such designs.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj committed:
drivers: lora: sx1276: select PA output by dts property

PA selection mainly depends on board design. So it looks like
device-tree is a better mechanism than Kconfig in this case. Use string
property with two possible values: "rfo" and "pa-boost".

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj committed:
drivers: lora: sx1276: support TCXO power control

Add support for TCXO power control using GPIO pin. Some boards
(including B-L072Z-LRWAN1 already supported in Zephyr) need delay
applied after powering on TCXO, so add device-tree property allowing to
configure that as well. Cache information about TCXO power status, so
subsequent requests to enable it will not result in unnecessary delays.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

KwonTae-young commented:
> There are several antenna path designs on SX1276 compatible boards in the wild. B-L072Z-LRWAN1 board has dedicated pins for RX enable and TX enable.

After this PR is merged, is SX1276 support of `b_l072z_lrwan1` board planned?
I am interested in the SX1276 support of `b_l072z_lrwan1`.

mniestroj commented:
>After this PR is merged, is SX1276 support of b_l072z_lrwan1 board planned?
I am interested in the SX1276 support of b_l072z_lrwan1.

Yes, I will add support for this board. Apart from changes in the MR, I need to have working Rtc* functions. This is now implemented using counter driver. I've got a commit to pass counter driver via dts, however it fails on sanitychecks due to missing counter driver for platforms on which tests are running. I do not think this is worth investing time, because @Mani-Sadhasivam plans to migrate to k_timer API instead. Which is the way to go in my opinion.

andysan reviewed (commented):
Any chance you could add support for a 3-way switch that selects between RX/TX_RFO/TX_PA? That seems to be what the Murata ABZ-078 uses. The reference implementation checks if PA boost has been enabled when enabling the TX path and selects the appropriate TX output in that case. This would likely have knock-on effect on antenna selection at some point.

mniestroj commented on drivers/lora/sx1276.c:
```diff
@@ -126,7 +132,9 @@ void SX1276SetAntSwLowPower(bool status)
 {
-	/* TODO */
+#if DT_INST_NODE_HAS_PROP(0, tcxo_power_gpios)
+	gpio_pin_set(dev_data.tcxo_power, GPIO_TCXO_POWER_PIN, state ? 1 : 0);
```
It is actually a good idea to include something like `tcxo-power-startup-delay` DT property.

mniestroj replied:
What I can see is that we do not provide `GetWakeupTime` callback as well. So it would be a good idea to start doing it, together with the introduction of above DT property.

andysan replied:
I noticed that as well after posting the comments. Another think I noticed (I started implementing pretty much the same changeset before finding you pull request) was that the reference LoRaWAN implementation turns on the TCXO in from the reset handler. I don’t know if that makes sense, but it’s probably a good idea to program a known state both in the TCXO and the antenna switch.

mniestroj replied:
I think that TCXO and antenna switch have known state already, because all pins are initialized with GPIO_OUTPUT_INACTIVE. However I agree with `SX1276Reset` case, we should do the same as reference implementation (power on TCXO before asserting reset pin).

andysan replied:
> > Any chance you could add support for a 3-way switch that selects between RX/TX_RFO/TX_PA? That seems to be what the Murata ABZ-078 uses. The reference implementation checks if PA boost has been enabled when enabling the TX path and selects the appropriate TX output in that case. This would likely have knock-on effect on antenna selection at some point.
> 
> So far sx1276 driver allows to select between RFO and PA_BOOST output in build time only. With that restriction in mind we do not need to have 3-way switch - 2-way switch is enough (either RX/TX_RFO or RX/TX_PA_BOOST).
> 
> If we would like to select between RFO and PA_BOOST in runtime, then we need to rework functionality selected by `power-amplifier-output` (or `CONFIG_PA_RFO_PIN` in master branch). @andysan Are there actually some use cases when you want to select between RFO and PA_BOOST in runtime? Have you some in mind?

My understanding is RFO path is significantly more power efficient than the PA boost path for the output powers where they overlap. I.e., if you are implementing a LoRa-WAN device that dynamically adjusts the output power (I believe this is pretty common), you'd get much better battery life if you disable PA boost when it isn't needed. Common platforms like the B-L072Z (Murata-based) hae a 3-way switch. The reference firmware only enables PA boost when the output power is higher than 14dB.

mniestroj commented:
>Any chance you could add support for a 3-way switch that selects between RX/TX_RFO/TX_PA? That seems to be what the Murata ABZ-078 uses. The reference implementation checks if PA boost has been enabled when enabling the TX path and selects the appropriate TX output in that case. This would likely have knock-on effect on antenna selection at some point.

So far sx1276 driver allows to select between RFO and PA_BOOST output in build time only. With that restriction in mind we do not need to have 3-way switch - 2-way switch is enough (either RX/TX_RFO or RX/TX_PA_BOOST).

If we would like to select between RFO and PA_BOOST in runtime, then we need to rework functionality selected by `power-amplifier-output` (or `CONFIG_PA_RFO_PIN` in master branch). @andysan Are there actually some use cases when you want to select between RFO and PA_BOOST in runtime? Have you some in mind?

andysan commented:
> > Any chance you could add support for a 3-way switch that selects between RX/TX_RFO/TX_PA? That seems to be what the Murata ABZ-078 uses. The reference implementation checks if PA boost has been enabled when enabling the TX path and selects the appropriate TX output in that case. This would likely have knock-on effect on antenna selection at some point.
> 
> So far sx1276 driver allows to select between RFO and PA_BOOST output in build time only. With that restriction in mind we do not need to have 3-way switch - 2-way switch is enough (either RX/TX_RFO or RX/TX_PA_BOOST).
> 
> If we would like to select between RFO and PA_BOOST in runtime, then we need to rework functionality selected by `power-amplifier-output` (or `CONFIG_PA_RFO_PIN` in master branch). @andysan Are there actually some use cases when you want to select between RFO and PA_BOOST in runtime? Have you some in mind?

My understanding is RFO path is significantly more power efficient than the PA boost path for the output powers where they overlap. I.e., if you are implementing a LoRa-WAN device that dynamically adjusts the output power (I believe this is pretty common), you'd get much better battery life if you disable PA boost when it isn't needed. Common platforms like the B-L072Z (Murata-based) hae a 3-way switch. The reference firmware only enables PA boost when the output power is higher than 14dB.

mniestroj commented:
>My understanding is RFO path is significantly more power efficient than the PA boost path for the output powers where they overlap. I.e., if you are implementing a LoRa-WAN device that dynamically adjusts the output power (I believe this is pretty common), you'd get much better battery life if you disable PA boost when it isn't needed. Common platforms like the B-L072Z (Murata-based) have a 3-way switch to support this. The reference firmware only enables PA boost when the output power is higher than 14dB.

My idea is to drop `antenna-rx-enable-gpios` and `antenna-tx-enable-gpios` and introduce:
- `rfi-enable-gpios`,
- `rfo-enable-gpios`,
- `pa-boost-enable-gpios`.
Additionally `power-amplifier-output` would be changed to `required: false`, because we know which one to enable based on `rfo-enable-gpios` and `pa-boost-enable-gpios`.

One case when we still need `power-amplifier-output` is when there is neither `rfo-enable-gpios` nor `pa-boost-enable-gpios` configured. This situation can be guarded by `BUILD_ASSERT()` with a proper warning about missing `rfo-enable-gpios`, `pa-boost-enable-gpios` and `power-amplifier-output` all at the same time.

mniestroj commented:
@andysan I think that all your comments have been addressed in pushed changes. Let me know if there is still something to improve :) Thanks for feedback so far!

andysan commented:
> @andysan I think that all your comments have been addressed in pushed changes. Let me know if there is still something to improve :) Thanks for feedback so far!

Thanks! I think you covered all of it. I tested the changes on a B-L072Z-LRWAN1, which has a 3-way switch, and they seem to work as expected.

andysan committed:
drivers: lora: sx1276: fix RFO TX power configuration

The PA_CONFIG register is currently not setup correctly for the RFO
path. The biggest problem is that the output power is incorrectly set
1dBm higher than requested. Additionally, the lower power levels are
not configured properly since the max power field in PA_CONFIG is
always configured for an output power between 0 and 15dBm.

To support lower than 0dBm, adjust the max power field in the
PA_CONFIG register to shift the output power range to -4--11 dBm when
the requested power is 0 or lower.

Signed-off-by: Andreas Sandberg <andreas@sandberg.pp.se>
[Marcin: rebase on master]
Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

Mani-Sadhasivam commented on drivers/lora/sx1276.c:
```diff
@@ -103,6 +113,10 @@ struct sx1276_data {
 #if DT_INST_NODE_HAS_PROP(0, rfo_enable_gpios) &&	\
 	DT_INST_NODE_HAS_PROP(0, pa_boost_enable_gpios)
 	u8_t tx_power;
+#endif
```
Please don't add this condition in middle. Add it after the previous `#endif`

mniestroj replied:
I'm not sure what you mean here.
```
#if DT_INST_NODE_HAS_PROP(0, rfi_enable_gpios)
	struct device *rfi_enable;
#endif
#if DT_INST_NODE_HAS_PROP(0, rfo_enable_gpios)
	struct device *rfo_enable;
#endif
#if DT_INST_NODE_HAS_PROP(0, pa_boost_enable_gpios)
	struct device *pa_boost_enable;
#endif
#if DT_INST_NODE_HAS_PROP(0, rfo_enable_gpios) &&	\
	DT_INST_NODE_HAS_PROP(0, pa_boost_enable_gpios)
	u8_t tx_power;
#endif
#if DT_INST_NODE_HAS_PROP(0, tcxo_power_gpios)
	struct device *tcxo_power;
	bool tcxo_power_enabled;
#endif
```
This is the result after change that you commented. How would you change it?

Mani-Sadhasivam replied:
The resulting code looks OK but I was concerned about the diff. If you look into your diff, you'll notice that you've added `#endif` and then `#ifdef`.

Mani-Sadhasivam replied:
You need to look into `drivers: lora: sx1276: support TCXO power control` commit to understand what I'm referring to.

mniestroj replied:
Understood. This is how diff algorithm works. We don't have influence which `#endif` is added in diff :)

Mani-Sadhasivam commented on drivers/lora/sx1276.c:
```diff
@@ -160,7 +174,25 @@ void SX1276SetAntSwLowPower(bool status)
+		return;
+	}
+
+	dev_data.tcxo_power_enabled = enable;
```
Move this statement below if/else condition

mniestroj replied:
Ok

Mani-Sadhasivam reviewed (changes requested):
Overall this PR looks good. I've left some comments inline. I'd like to get rid of `power-amplifier-output` property in favor of RFO/RFI/PA-BOOST gpios. It is still used in `96b_wistrio` but converting that would be trivial. Let me know if you need any help for that.

I'll test the next version on my setup.

mniestroj commented on drivers/lora/sx1276.c:
```diff
@@ -313,29 +313,19 @@ void SX1276SetRfTxPower(int8_t power)
 
 	if (power > 17) {
 		pa_dac = (pa_dac & RF_PADAC_20DBM_MASK) | RF_PADAC_20DBM_ON;
+		power -= 3;
```
In case of 20DBM_ON there was `(power - 5)`, in the other case it was `(power - 2)`. So here comes the difference. But `power -= 3` can be confusing, I agree.

mniestroj commented on drivers/lora/sx1276.c:
```diff
@@ -84,9 +127,35 @@ void RtcStopAlarm(void)
 void SX1276SetAntSwLowPower(bool status)
 {
-	/* TODO */
+	if (status) {
```
What about:
- changing `s/status/low_power/`
- adding `/* force inactive (low power) state of all antenna paths */` in `low_power`
- adding `/* rely on SX1276SetAntSw() to configure proper antenna path */` in `!low_power`

mniestroj commented on drivers/lora/sx1276.c:
```diff
@@ -313,29 +313,19 @@ void SX1276SetRfTxPower(int8_t power)
-
-		pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
-			     ((power - 2) & 0x0F);
+	if (power > 17) {
```
I agree that @andysan patch improves the code a lot.

mniestroj commented:
> Upon further testing, I discovered that the RFO TX power configuration in the original code is incorrect, the output power is off by 1dBm. I fixed the original problem and added support for down to -4dBm by setting the max power field in the PA_CONFIG register. You can find the change here: https://github.com/andysan/zephyr/commits/sx1276_power_fixes
> 
> It's a fairly trivial fix, so feel free to incorporate it into your cleanup commit or any way you find appropriate.

I like your implementation. I prefer to use it as first patch in PR, so we don't modify the same thing (power output logic) twice. Also I think we should use `s8_t` instead of `int8_t` types. I've modified your commit and put: https://github.com/mniestroj/zephyr/commits/sx1276-dts-props-v2. Please check whether it is still fine :)

mniestroj commented:
> Overall this PR looks good. I've left some comments inline. I'd like to get rid of `power-amplifier-output` property in favor of RFO/RFI/PA-BOOST gpios. It is still used in `96b_wistrio` but converting that would be trivial. Let me know if you need any help for that.
> 
> I'll test the next version on my setup.

I think it is still possible that neither RFO nor PA-BOOST will be configured:
- two separate antennas for TX and RX (e.g. you want to have RX active 100% time - this is however not supported in loramac-node)
- TX only with antenna always "enabled"

But maybe those are too theoretical ;)

andysan commented:
> > Upon further testing, I discovered that the RFO TX power configuration in the original code is incorrect, the output power is off by 1dBm. I fixed the original problem and added support for down to -4dBm by setting the max power field in the PA_CONFIG register. You can find the change here: https://github.com/andysan/zephyr/commits/sx1276_power_fixes
> > It's a fairly trivial fix, so feel free to incorporate it into your cleanup commit or any way you find appropriate.
> 
> I like your implementation. I prefer to use it as first patch in PR, so we don't modify the same thing (power output logic) twice. Also I think we should use `s8_t` instead of `int8_t` types. I've modified your commit and put: https://github.com/mniestroj/zephyr/commits/sx1276-dts-props-v2. Please check whether it is still fine :)

Thanks for cleaning up the int types. I had a look at the updated patch set (only the power selection bits) and they look OK. However, I did realise that we mask out every single bit in the PACONFIG register, so we could just set it to 0 instead of reading it.

andysan commented:
> > Overall this PR looks good. I've left some comments inline. I'd like to get rid of `power-amplifier-output` property in favor of RFO/RFI/PA-BOOST gpios. It is still used in `96b_wistrio` but converting that would be trivial. Let me know if you need any help for that.
> > I'll test the next version on my setup.
> 
> I think it is still possible that neither RFO nor PA-BOOST will be configured:
> 
> * two separate antennas for TX and RX (e.g. you want to have RX active 100% time - this is however not supported in loramac-node)
> * TX only with antenna always "enabled"
> 
> But maybe those are too theoretical ;)

The datasheet actually specifies that you can connect RFO and RFI using passive components to avoid an RF switch, so it's definitely possible. I haven't seen such devices in the wild, but I haven't been looking that hard.

mniestroj commented:
> > > Upon further testing, I discovered that the RFO TX power configuration in the original code is incorrect, the output power is off by 1dBm. I fixed the original problem and added support for down to -4dBm by setting the max power field in the PA_CONFIG register. You can find the change here: https://github.com/andysan/zephyr/commits/sx1276_power_fixes
> > > It's a fairly trivial fix, so feel free to incorporate it into your cleanup commit or any way you find appropriate.
> > 
> > 
> > I like your implementation. I prefer to use it as first patch in PR, so we don't modify the same thing (power output logic) twice. Also I think we should use `s8_t` instead of `int8_t` types. I've modified your commit and put: https://github.com/mniestroj/zephyr/commits/sx1276-dts-props-v2. Please check whether it is still fine :)
> 
> Thanks for cleaning up the int types. I had a look at the updated patch set (only the power selection bits) and they look OK. However, I did realise that we mask out every single bit in the PACONFIG register, so we could just set it to 0 instead of reading it.

Good catch. Then I will remove:
```
	pa_config &= RF_PACONFIG_MAX_POWER_MASK;
	pa_config &= RF_PACONFIG_OUTPUTPOWER_MASK;
	pa_config &= RF_PACONFIG_PASELECT_MASK;
```
in the updated PR.

mniestroj commented:
Should we remove reading PA_DAC as well? Datasheet provides default value (0x10) for "reverved" field, which is the only one we do not overwrite now.

mniestroj committed:
drivers: lora: sx1276: stop reading PA_CONFIG

All fields in PA_CONFIG register are set explicitly, so there is no need
to read this register first.

Suggested-by: Andreas Sandberg <andreas@sandberg.pp.se>
Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

andysan commented:
> Should we remove reading PA_DAC as well? Datasheet provides default value (0x10) for "reverved" field, which is the only one we do not overwrite now.

Good question. I thought about that and decided not to suggest it. The main reasoning was that since it is a value with an unknown meaning, I didn't trust the default in the datasheet.

mniestroj commented on drivers/lora/sx1276.c:
```diff
@@ -22,15 +23,44 @@ LOG_MODULE_REGISTER(sx1276);
+#define SX1276_PA_OUTPUT(power)		SX1276_PA_RFO
+#elif DT_INST_NODE_HAS_PROP(0, pa_boost_enable_gpios)
+#define SX1276_PA_OUTPUT(power)		SX1276_PA_BOOST
+#elif DT_INST_NODE_HAS_PROP(0, power_amplifier_output)
```
There is a possibility that another board might use it, so I opt for keeping this functionality.

Mani-Sadhasivam commented on drivers/lora/sx1276.c:
```diff
@@ -22,6 +22,15 @@ LOG_MODULE_REGISTER(sx1276);
+ * Those macros must be in sync with 'power-amplifier-output' dts property.
+ */
+#define SX1276_PA_RFO		0
+#define SX1276_PA_BOOST		1
```
Align this to the below `SX1276_REG_PA_CONFIG` define

Mani-Sadhasivam commented on drivers/lora/sx1276.c:
```diff
@@ -38,6 +38,16 @@ LOG_MODULE_REGISTER(sx1276);
+#define GPIO_TCXO_POWER_PIN	DT_INST_GPIO_PIN(0, tcxo_power_gpios)
+#define GPIO_TCXO_POWER_FLAGS	DT_INST_GPIO_FLAGS(0, tcxo_power_gpios)
+
+#if DT_INST_NODE_HAS_PROP(0, tcxo_power_startup_delay_ms)
```
Just for my understanding. From where this delay requirement comes in? Board design or crystal osc?

andysan replied:
A TCXO (TC stands for Temperature Controlled) contains a tiny oven that keeps the crystal at a well-defined temperature. Due to the physics of the device, it'll take some time between turning on power and the device reaching its target temperature. The frequency of the crystal is only well defined once it reaches its target temperature.

The trick here is that you want the delay to be enough to heat the crystal, but not too long since that wastes battery.

Mani-Sadhasivam commented on drivers/lora/sx1276.c:
```diff
@@ -296,47 +309,38 @@ void SX1276SetRfTxPower(int8_t power)
-		pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
-			     ((power - 2) & 0x0F);
+		/* Set the power range to -4.2 -- 10.8+0.6*0 dBm */
+		pa_config |= 0 << SX1276_PA_CONFIG_MAX_POWER_SHIFT;
```
this is a nop, please remove.

andysan replied:
We probably want to keep the comment in that case since the MAX_POWER field affects how the power is calculated. The reason I added the line in the first place was as documentation.

Mani-Sadhasivam replied:
Yes, let's keep the comment.

Mani-Sadhasivam reviewed (changes requested):
Few nitpicks. Once fixed, rest LGTM.

Mani-Sadhasivam reviewed (approved):
LGTM

andysan commented on drivers/lora/sx1276.c:
```diff
@@ -52,35 +108,143 @@ static const struct sx1276_dio sx1276_dios[] = { SX1276_DIO_GPIO_INIT(0) };
+#endif
+#if DT_INST_NODE_HAS_PROP(0, rfo_enable_gpios) &&	\
+	DT_INST_NODE_HAS_PROP(0, pa_boost_enable_gpios)
+	u8_t tx_power;
```
This should be uint8_t nowadays.

mniestroj replied:
Good catch! ;) Fixed!

andysan reviewed (changes requested):
I think you missed one of the old integer types. Looks good other than that.

andysan reviewed (approved):
LGTM. Thanks!

andysan commented:
@carlescufi: This PR should be ready to be merged now.
